zookeeper-docs/src/main/resources/markdown/zookeeperOracleQuorums.md

```diff
@@ -0,0 +1,128 @@
+<!--
+Copyright 2002-2004 The Apache Software Foundation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+//-->
+
+# Introduction to Oracle Quorum
+
+The introduction to Oracle Quorum increases the availability of a cluster of 2 ZooKeeper instances with a failure detector
+ as known as the Oracle. The Oracle is designed to grant the permission to the instance which is the only remaining instance
+in a 2-node configuration when the other instance is identified as faulty by the fail detector, the Oracle.
+
+## The implementation of the Oracle
```

Comment:
Do we have a canonical reference implementation of such an Oracle somewhere? After reading the doc, I only know an Oracle is a failure detector but I am clueless on how to implement one for myself, or where to get one if I want to try this feature.

Reply:
In the previous deployed context, the oracle is a text file which contains either 0 or 1. 0 means the other instance is not faulty, and vice versa. Due to the heterogeneous deployed contexts, the ways of implementation of an Oracle are not defined. An example could be a hardware device which could distinguishes whether the machine is power on or not. Based the information, the device could update the file on the instance to 0 or 1. Another example could be another process within the same machine as the zookeeper instance constantly check the other instance is alive or not, for instance, by pinging the other node , and bases the information to update the oracle file. In the case for hard to find two dedicated hardware, we could use docker to stimulate a two-node systems (https://github.com/chingchan1996/Docker_ZooKeeper), and manually cause failures and update the oracle files.

Reply:
some how I cannot reply under the comment for the unit test one.
Could you give me more clues on which examples I should look up? 
If it needs many changes, I think we should do it in another PR. This PR is kind of big from my point view.

Reply:
You could do a search of "@ParameterizedTest" under zookeeper-server source folder and checkout the samples.

For this specific case, I think we'd have to refactor the existing methods and remove the boolean withObserver parameter and replace it with a new abstraction that can encode more state (with / o observers and oracles). I don't consider this is a blocker for this PR, so it's fine if you decide to do it as a separate commit. The code just smell so badly with crowded boolean parameters on methods.

Reply:
> In the previous deployed context, the oracle is a text file which contains either 0 or 1. 0 means the other instance is not faulty, and vice versa. Due to the heterogeneous deployed contexts, the ways of implementation of an Oracle are not defined. An example could be a hardware device which could distinguishes whether the machine is power on or not. Based the information, the device could update the file on the instance to 0 or 1. Another example could be another process within the same machine as the zookeeper instance constantly check the other instance is alive or not, for instance, by pinging the other node , and bases the information to update the oracle file. In the case for hard to find two dedicated hardware, we could use docker to stimulate a two-node systems (https://github.com/chingchan1996/Docker_ZooKeeper), and manually cause failures and update the oracle files.

I think these info (or in a rephrased form) are valuable to be added in the doc.

Reply:
For the UT,
I will first have another branch to handle it. If the changes are ok, then we merge to the current branch.

For the deployment info, I will organize them and update them to the doc. Thanks for the comments :)